enhancements/installer/cluster-api-based-installations.md

```diff
@@ -0,0 +1,347 @@
+2. (optional) The cluster creator runs `openshift-install create manifests`
+3. (optional) The cluster creator edits the newly created CAPI manifests.
+4. The cluster creator runs `openshift-install create cluster`
+5. `openshift-install` extracts binaries for kube-apiserver, etcd, CAPI infrastructure provider & cloud CAPI provider to the install dir
+6. `openshift-install` using `envtest` initializes a control plane locally on the Installer host
+7. `openshift-install` execs the CAPI infrastructure and cloud provider as subprocesses, pointing them to the local control plane
+8. `openshift-install` applies the CAPI manifests to the control plane
+9. The CAPI controllers provision cluster infrastructure based on the manifests
+10. `openshift-install` monitors the status of the local manifests as they are applied
+11. If the statuses are as expected, infrastrucutre has been provisioned and installation continues with the normal flow.
+
+In the case of an error in the final step, the Installer will bubble up resources with non-expected statuses.
+
+#### Variation and form factor considerations [optional]
+
+
+### API Extensions
+
+As a result of this enhancement, API Extensions will only be used locally by the 
+local Installer control plane--they will not, until future work, be pivoted to the cluster.
+The following are CRD references for Installer-supported platforms:
+
+- AWS: [infrastructure.cluster.x-k8s.io/v1beta2][awsCRD]
+- Azure/AzureStack: [infrastructure.cluster.x-k8s.io/v1beta1][azureCRD]
+- GCP: [infrastructure.cluster.x-k8s.io/v1beta1][gcpCRD]
+- IBMCloud/PowerVS: [infrastructure.cluster.x-k8s.io/v1beta2][ibmCRD]
+- Nutanix: [infrastructure.cluster.x-k8s.io/v1beta1][nutanixCRD]
+- OpenStack: [infrastructure.cluster.x-k8s.io/v1alpha7][openstackCRD]
+- vSphere: [infrastructure.cluster.x-k8s.io/v1beta1][vsphereCRD]
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Overview
+
+In a typical CAPI installation, manifests indicating the desired cluster configuration are applied to a
+management cluster. In order to keep `openshift-install` free of new external runtime dependencies,
+the management-cluster dependencies will be [embedded][embed] into the `openshift-install` binary,
+extracted at runtime, and cleaned up afterward.  This approach is similar to what we have been using
+for Terraform.
+
+In its current configuration, the Installer embds the Terraform and cloud-specific provider binaries
+within the `openshift-install` binary and extracts them at runtime. The Installer produces the Terraform
+configuration files and invokes Terraform using the `tf-exec` library.
+
+![alt terraform diagram](terraform_embedded.jpg)
+
+
+We can follow a similar pattern to run CAPI controllers locally on the Installer host. In addition
+to the CAPI controller binaries, `kube-apiserver` and `etcd` are embedded in order to run a local
+control plane, orchestrated with `envtest`.
```

Comment:
There was an idea in the past to have the terraform binaries in the release image instead of embedded in the Installer binary. We can revisit that idea if there is no case in which the Installer is used without a release image (I imagine most customers get the installer *from* the release payload).

Reply:
IIRC, when the agent team looked into adding the code needed to download a container image and extract something out of it the found that it made the dependencies of the installer impossible to manage. @zaneb am I remembering that correctly?

Reply:
This is a great suggestion, and I will add it. I am convinced that we should pull the providers from the release image, but I am not certain whether that should be at build or run time.

I have implemented the release image extraction previously. Here is [a link to the changes in the go.mod](https://github.com/openshift/installer/pull/6599/commits/5f3e07883ffe1b7bb01f69965b38d064b110eb8d).

Reply:
The installer binary can also be downloaded by itself here: https://console.redhat.com/openshift/downloads

Reply:
There are trade-offs with having them in the payload vs. just having them on quay.io vs. building them into the binary. It's more convenient in some ways to put them in the payload, but will also have an impact on the payload size, and we're trying to shrink (or at least carefully manage) that. I know we have a goal of eventually pivoting the CAPI resources into a cluster, but that won't apply for every cluster, which is another reason to not include them in the payload all the time. It would be helpful to understand if it would be possible to deliver the CAPI providers via OLM, download the providers from the installer, and also pivot into a cluster later, when the user wants it. That lets us keep the payload size smaller, starts CAPI out as one of our first "platform operators" as part of the "composable OpenShift" goal. I wouldn't block all progress on this over the question, but it would be good to spend a little time thinking it through to see what's possible.

Reply:
> IIRC, when the agent team looked into adding the code needed to download a container image and extract something out of it the found that it made the dependencies of the installer impossible to manage. @zaneb am I remembering that correctly?

Correct. Naively importing the oc code that does this pulls in all of k8s/k8s.
In theory it is [possible](https://issues.redhat.com/browse/AGENT-307) to reimplement the extraction code using ~only the Docker client lib, but it is [haaaard](https://github.com/zaneb/iso_extractor).
Shelling out to `oc` is easy and what the agent installer actually does, but this would impose a new requirement for IPI users.

Extracting the binaries from the release payload at runtime would be the ideal solution in my view, but Service Delivery have asked us not to do that because any network calls the installer makes get a lot of scrutiny from a security compliance perspective. (I am sceptical that contacting the registry is a big problem given that the cluster itself has to do the same, but it was @JoelSpeed who actually received this feedback so he might have more details.)

Extracting the binaries from the release payload at build time makes builds faster but does not solve the CI dependency issue nor the binary size issue.

Reply:
Do we think this is a blocker for the first iteration of this proposal, or is this something we can defer to a future enhancement?

Reply:
I do think we need a solution for how we're going to actually deliver this end-to-end, and that includes understanding how we're going to build for binaries for users as well as in CI so it can be tested.

Since introducing code to manage downloading images makes building the installer impractical at all, and embedding the CAPI binaries makes CI challenging, maybe the installer could embed oc instead of all of the CAPI binaries?

Reply:
> Do we think this is a blocker for the first iteration of this proposal, or is this something we can defer to a future enhancement?

The only technical blocker is that we need to figure out how to handle/build the kube-apiserver and etcd dependencies. The OpenShift Build System (OSBS) cannot download dependencies, which is how these two binaries are being obtained at the moment. This pattern (downloading) works in CI, but not when building the product.

> embedding the CAPI binaries makes CI challenging

AFAIK embedding the binaries is fine for CI and ART. It's the same thing we do for Terraform. Vendoring the dependencies into the installer and then building them with the installer works but has significant drawbacks:

1. slower builds
2. changes need to be committed to the provider then vendored to the installer

The first drawback is already a significant pain point for the installer, while the second will become more significant with this enhancement. I wouldn't call them blockers, but are high priority in my opinion. I will add further discussion of design ideas for how to utilize the release image for the build time and runtime dependency resolution, but  how to get etcd and kube-apiserver is an open question for me so we need some help there.

Reply:
I'd be fine with that if CI could do it.

Reply:
>  anything else?

Not a negative, but we would need to account for the local developer experience: create a Makefile or script that will copy the binaries from the release image to the developer's local machine to build the installer.